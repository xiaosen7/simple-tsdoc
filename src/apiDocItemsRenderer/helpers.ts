import * as types from "../types";

export function callCustomTagRenderAction(
  customTag: types.CustomTag,
  baseApi: types.RenderActionApi,
  tagPayload: types.Annotation["tags"][number] | null
) {
  const renderAction = customTag.renderAction;

  if (typeof renderAction === "function") {
    renderAction({
      ...baseApi,
      payload: tagPayload,
    });
  }
}
