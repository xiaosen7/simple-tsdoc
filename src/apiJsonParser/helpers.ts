import * as model from "@microsoft/api-extractor-model";
import * as types from "../types";
import * as tsdoc from "@microsoft/tsdoc";
import * as commentParser from "../commentParser";

const canVisitKinds = [
  model.ApiItemKind.Class,
  model.ApiItemKind.Function,
  model.ApiItemKind.Method,
  model.ApiItemKind.Property,
  model.ApiItemKind.Constructor,
  model.ApiItemKind.Namespace,
] as const;
type CanVisitKind = typeof canVisitKinds[number];
type AfterVisitKind = `After${CanVisitKind}`;
export type Visitor = Partial<
  Record<CanVisitKind | AfterVisitKind, (apiDocItem: types.ApiDocItem) => any>
>;

export function getAnnotation(apiItem: model.ApiDeclaredItem) {
  const tsdocComment: tsdoc.DocComment | undefined = apiItem.tsdocComment;

  let annotation = defaultAnnotation;
  if (tsdocComment) {
    annotation = commentParser.parse(tsdocComment.emitAsTsdoc());
  }

  return annotation;
}

export function traversePackage(
  apiPackage: model.ApiPackage,
  visitor: Visitor
) {
  traverseMembers(apiPackage.members, visitor);
}

function traverseCallback(apiItem: model.ApiItem, visitor: Visitor) {
  let apiDocItem!: types.ApiDocItem;
  let kind!: model.ApiItemKind;
  let canVisit = false;

  if (apiItem instanceof model.ApiDeclaredItem) {
    const annotation = getAnnotation(apiItem);
    kind = getKind(apiItem, annotation);

    if (canVisitKinds.includes(kind as any)) {
      apiDocItem = {
        annotation,
        kind,
        name: apiItem.displayName,
        apiItem,
      };

      canVisit = true;
    }
  }

  if (canVisit) {
    const fn = visitor[kind as CanVisitKind];
    fn && fn(apiDocItem);
  }

  if (apiItem.members) {
    traverseMembers(apiItem.members, visitor);
  }

  if (canVisit) {
    const after = visitor[getAfterKind(kind as CanVisitKind)];
    after && after(apiDocItem);
  }
}

function traverseMembers(members: readonly model.ApiItem[], visitor: Visitor) {
  if (members?.length) {
    members.forEach((apiItem) => traverseCallback(apiItem, visitor));
  }
}
export const defaultAnnotation: types.Annotation = {
  tags: [],
  deprecated: false,
};

export function isFunctionKind(
  apiItem: model.ApiItem,
  annotation: types.Annotation
) {
  return (
    apiItem.kind === model.ApiItemKind.Function ||
    (apiItem.kind === model.ApiItemKind.Variable &&
      annotation.modifierTags?.includes("@function"))
  );
}

export function isClassKind(
  apiItem: model.ApiItem,
  annotation: types.Annotation
) {
  return (
    apiItem.kind === model.ApiItemKind.Class ||
    (apiItem.kind === model.ApiItemKind.Variable &&
      annotation.modifierTags?.includes("@constructor"))
  );
}

export function getKind(
  apiItem: model.ApiItem,
  annotation: types.Annotation
): model.ApiItemKind {
  if (isFunctionKind(apiItem, annotation)) {
    return model.ApiItemKind.Function;
  }

  if (isClassKind(apiItem, annotation)) {
    return model.ApiItemKind.Class;
  }

  return apiItem.kind;
}

export function getAfterKind(kind: CanVisitKind): AfterVisitKind {
  return `After${kind}` as const;
}
