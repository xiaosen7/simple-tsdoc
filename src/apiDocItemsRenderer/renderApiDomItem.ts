import { ApiDocItem, Renderer, RenderAction, RenderActionApi } from "../types";
import * as utils from "../utils";
import { customTagsMap } from "../customTags";
import { callCustomTagRenderAction } from "./helpers";
import * as model from "@microsoft/api-extractor-model";
import { formatDocNode } from "../shared";
import { I18n } from "i18n";

const defaultTagRenderActions: Record<string, RenderAction> = {
  "@example": (api) => {
    const { payload, i18n } = api;
    if (utils.isObject(payload)) {
      api.appendMd(`\n${i18n.__("examples")}：\n` + payload.description + `\n`);
    }
  },
};

const renderTypeParameters: RenderAction = (api) => {
  const {
    apiDocItem: { apiItem },
    appendMd,
    i18n,
  } = api;

  if (
    apiItem instanceof model.ApiFunction ||
    apiItem instanceof model.ApiMethod
  ) {
    if (apiItem.parameters.length === 0) {
      return;
    }

    const parametersContent = apiItem.parameters
      .map((parameter) => {
        const { name, isOptional } = parameter;
        const parameterType = parameter.parameterTypeExcerpt.text;
        let description;
        if (parameter.tsdocParamBlock) {
          description = formatDocNode(parameter.tsdocParamBlock.content);
        }

        return `\n- ${name}(${parameterType}, ${
          isOptional ? i18n.__("optional") : i18n.__("necessary")
        }) ${description ?? ""}\n`;
      })
      .join("\n");

    appendMd(`\n${i18n.__("parameters")}：\n${parametersContent}\n`);
  }
};

const renderParams: RenderAction = (api) => {
  const { annotation, i18n } = api;

  api.appendMd(
    `\n${i18n.__("parameters")}：\n${annotation
      .params!.map(({ name, description }) => {
        return `- ${name} ${description}`;
      })
      .join("\n")}\n`
  );
};

const renderDescription: RenderAction = (api) => {
  api.appendMd(
    `\n${api.i18n.__("description")}：${api.annotation.description}\n`
  );
};

const renderReturns: RenderAction = (api) => {
  api.appendMd(`\n${api.i18n.__("returns")}：${api.annotation.returns}\n`);
};

const renderDeprecated: RenderAction = (api) => {
  api.appendMd(`\n${api.i18n.__("returns")}：${api.annotation.returns}\n`);
};

const renderModifierTags: RenderAction = (api) => {
  api.appendMd(`\n${api.payload}\n`);
};

const renderKind: RenderAction = (api) => {
  api.appendMd(`\n${api.i18n.__("kind")}：${api.apiDocItem.kind.toString()}\n`);
};

export function renderApiDomItem(
  apiDocItem: ApiDocItem,
  renderers: Renderer[],
  titleLevel: number,
  i18n: I18n,
  appendMd?: (content: string) => any
) {
  const { annotation, apiItem } = apiDocItem;
  const baseApi: RenderActionApi = {
    apiDocItem,
    payload: null,
    annotation,
    appendMd: null as any,
    i18n,
  };

  if (appendMd) {
    // means this is property apiDocItem
    appendMd(`\n${"#".repeat(titleLevel)} ${apiDocItem.name}\n`);

    baseApi.appendMd = appendMd;
  } else {
    let pendingMd = [`\n${"#".repeat(titleLevel)} ${apiDocItem.name}\n`];

    renderers.push({
      render: () => pendingMd.join("\n"),
      apiDocItem,
    });

    baseApi.appendMd = (content: string) => pendingMd.push(content);
  }

  // render description
  if (utils.isValidString(annotation.description)) {
    renderDescription(baseApi);
  }

  // render kind
  renderKind(baseApi);

  // render tags
  annotation.tags.forEach((tag) => {
    if (customTagsMap.has(tag.name)) {
      callCustomTagRenderAction(customTagsMap.get(tag.name)!, baseApi, tag);
    } else {
      if (tag.name in defaultTagRenderActions) {
        defaultTagRenderActions[tag.name]({
          ...baseApi,
          payload: tag,
        });
      }
    }
  });

  // render modifierTags
  if (utils.isValidArray(annotation.modifierTags)) {
    annotation.modifierTags!.forEach((tagName) => {
      if (customTagsMap.has(tagName)) {
        callCustomTagRenderAction(customTagsMap.get(tagName)!, baseApi, null);
      } else {
        renderModifierTags({
          ...baseApi,
          payload: tagName,
        });
      }
    });
  }

  // render params
  if (
    apiItem instanceof model.ApiFunction ||
    apiItem instanceof model.ApiMethod
  ) {
    renderTypeParameters(baseApi);
  } else if (utils.isValidArray(annotation.params)) {
    renderParams(baseApi);
  }

  // render returns
  if (utils.isValidString(annotation.returns)) {
    renderReturns(baseApi);
  }

  // render deprecated
  if (utils.isValidString(annotation.deprecated)) {
    renderDeprecated(baseApi);
  }

  // render properties
  if (utils.isValidArray(apiDocItem.properties)) {
    apiDocItem.properties?.forEach((propertyApiDocItem) => {
      renderApiDomItem(
        propertyApiDocItem,
        renderers,
        titleLevel + 1,
        i18n,
        baseApi.appendMd
      );
    });
  }
}
