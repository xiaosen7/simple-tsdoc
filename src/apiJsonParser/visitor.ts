import * as helpers from "./helpers";
import * as model from "@microsoft/api-extractor-model";
import { ApiDocItem } from "../types";
import * as utils from "../utils";

export function createVisitor(docItems: ApiDocItem[]): helpers.Visitor {
  const classApiDocItemStack: ApiDocItem[] = [];
  const namespaceApiDocItemStack: ApiDocItem[] = [];

  return {
    [model.ApiItemKind.Function](apiDocItem) {
      if (utils.peek(namespaceApiDocItemStack)) {
        pushIntoProperties(namespaceApiDocItemStack, apiDocItem);
      } else {
        docItems.push(apiDocItem);
      }
    },
    [model.ApiItemKind.Class](apiDocItem) {
      classApiDocItemStack.push(apiDocItem);
      if (utils.peek(namespaceApiDocItemStack)) {
        pushIntoProperties(namespaceApiDocItemStack, apiDocItem);
      } else {
        docItems.push(apiDocItem);
      }
    },
    [helpers.getAfterKind(model.ApiItemKind.Class)]() {
      classApiDocItemStack.pop();
    },
    [model.ApiItemKind.Property](apiDocItem) {
      pushIntoProperties(classApiDocItemStack, apiDocItem);
    },
    [model.ApiItemKind.Method](apiDocItem) {
      pushIntoProperties(classApiDocItemStack, apiDocItem);
    },
    [model.ApiItemKind.Constructor](apiDocItem) {
      pushIntoProperties(classApiDocItemStack, apiDocItem);
    },
    [model.ApiItemKind.Namespace](apiDocItem) {
      namespaceApiDocItemStack.push(apiDocItem);
      docItems.push(apiDocItem);
    },
    [helpers.getAfterKind(model.ApiItemKind.Namespace)]() {
      namespaceApiDocItemStack.pop();
    },
  };
}

function pushIntoProperties(parentStack: ApiDocItem[], apiDocItem: ApiDocItem) {
  const parent = utils.peek(parentStack);
  if (parent) {
    if (!parent.properties) {
      parent.properties = [];
    }

    parent.properties.push(apiDocItem);
  }
}
