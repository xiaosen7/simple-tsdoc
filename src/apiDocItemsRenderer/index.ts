import * as types from "../types";
import * as renderActions from "./renderApiDomItem";
import i18n from "i18n";
import { resolve } from "path";

export const locales = ["en", "ch"];
i18n.configure({
  locales,
  directory: resolve(__dirname, "..", "locales"),
});

export type Lang = typeof locales[number];

export interface RenderOptions {
  /**
   * The title level of per function.
   *
   * If set to 3, then a function title will be prefixed by '###' in generated markdown content.
   *
   * @default 3
   */
  functionTitleLevel?: number;

  /**
   * Set the language of rendering.
   * @default 'en'
   */
  lang: Lang;
}

export function render(
  apiDocItems: types.ApiDocItem[],
  options: RenderOptions = {
    lang: "en",
  }
) {
  const fullOptions: Required<RenderOptions> = Object.assign(
    {
      functionTitleLevel: 3,
      lang: "en",
    },
    options
  );

  i18n.setLocale(fullOptions.lang);

  const renderers: types.Renderer[] = [];

  apiDocItems.forEach((apiDocItem) =>
    renderActions.renderApiDomItem(
      apiDocItem,
      renderers,
      fullOptions.functionTitleLevel,
      i18n
    )
  );

  return renderers;
}
