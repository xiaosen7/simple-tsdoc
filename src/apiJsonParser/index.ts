import * as model from "@microsoft/api-extractor-model";
import { ApiDocItem } from "../types";
import { traversePackage } from "./helpers";
import { createVisitor } from "./visitor";

export function parse(apiJsonFilePath: string) {
  const apiModel: model.ApiModel = new model.ApiModel();
  const apiPackage: model.ApiPackage = apiModel.loadPackage(apiJsonFilePath);

  const docItems: ApiDocItem[] = [];
  const visitor = createVisitor(docItems);
  traversePackage(apiPackage, visitor);

  return docItems;
}
