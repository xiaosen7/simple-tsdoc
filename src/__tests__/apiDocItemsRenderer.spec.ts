import * as testUtils from "./testUtils";
import * as apiDocItemsRenderer from "../apiDocItemsRenderer";
import * as apiJsonParser from "../apiJsonParser";

test("apiDocItemsRenderer", () => {
  const inputFilePath = testUtils.resolveAssetPath(
    "api-extractor-demo.api.json"
  );
  const docItems = apiJsonParser.parse(inputFilePath);
  const renderers = apiDocItemsRenderer.render(docItems);
  expect(renderers.map(({ render }) => render())).toMatchSnapshot();
});
