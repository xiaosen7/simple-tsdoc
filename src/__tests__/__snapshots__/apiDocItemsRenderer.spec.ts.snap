// Vitest Snapshot v1

exports[`apiDocItemsRenderer 1`] = `
Map {
  "其他" => [
    {
      "apiDocItem": {
        "annotation": {
          "deprecated": false,
          "description": "",
          "modifierTags": [
            "@function",
          ],
          "namespace": "其他",
          "params": [],
          "returns": undefined,
          "tags": [],
        },
        "kind": "Function",
        "name": "arrowFunction",
      },
      "render": [Function],
    },
    {
      "apiDocItem": {
        "annotation": {
          "deprecated": false,
          "description": "",
          "modifierTags": [
            "@constructor",
          ],
          "namespace": "其他",
          "params": [],
          "returns": undefined,
          "tags": [],
        },
        "kind": "Class",
        "name": "createdClass",
        "properties": [],
      },
      "render": [Function],
    },
    {
      "apiDocItem": {
        "annotation": {
          "deprecated": false,
          "description": "",
          "modifierTags": [
            "@function",
          ],
          "namespace": "其他",
          "params": [],
          "returns": undefined,
          "tags": [],
        },
        "kind": "Function",
        "name": "createdFunction",
      },
      "render": [Function],
    },
    {
      "apiDocItem": {
        "annotation": {
          "deprecated": false,
          "description": "人这个类",
          "modifierTags": [],
          "namespace": "其他",
          "params": [],
          "returns": undefined,
          "tags": [],
        },
        "kind": "Class",
        "name": "Person",
        "properties": [
          {
            "annotation": {
              "deprecated": false,
              "description": "Constructs a new instance of the \`Person\` class",
              "modifierTags": [],
              "namespace": "Person",
              "params": [],
              "returns": undefined,
              "tags": [],
            },
            "kind": "Constructor",
            "name": "(constructor)",
          },
          {
            "annotation": {
              "deprecated": false,
              "namespace": "Person",
              "tags": [],
            },
            "kind": "Method",
            "name": "getName",
          },
          {
            "annotation": {
              "deprecated": false,
              "namespace": "Person",
              "tags": [],
            },
            "kind": "Property",
            "name": "name",
          },
        ],
      },
      "render": [Function],
    },
  ],
  "math" => [
    {
      "apiDocItem": {
        "annotation": {
          "deprecated": false,
          "description": "add 函数",
          "modifierTags": [
            "@public",
          ],
          "namespace": "math",
          "params": [],
          "returns": undefined,
          "tags": [],
        },
        "kind": "Function",
        "name": "add",
      },
      "render": [Function],
    },
  ],
  "foobar" => [
    {
      "apiDocItem": {
        "annotation": {
          "deprecated": false,
          "description": "foo函数",
          "modifierTags": [],
          "namespace": "foobar",
          "params": [
            {
              "description": "",
              "name": "name",
            },
            {
              "description": "",
              "name": "age",
            },
          ],
          "returns": "",
          "tags": [
            {
              "description": "\`\`\`ts
const res = foo('xiaoming', 18); // => xiaoming18
\`\`\`",
              "name": "@example",
            },
          ],
        },
        "kind": "Function",
        "name": "foo",
      },
      "render": [Function],
    },
  ],
}
`;
